src/vmm/src/irq_allocator.rs

```diff
@@ -0,0 +1,84 @@
+use std::fmt;
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// An irq allocator which gives next available irq.
+/// It is mainly used for non-legacy devices.
+// There are a few reserved irq's on x86_64. We just skip all the inital
+// reserved irq to make the implementaion simple. This could be later extended
+// to cater more complex scenario.
+#[derive(Debug)]
+pub struct IrqAllocator {
+    // Tracks the last allocated irq
+    last_used_irq: u32,
+    max: u32,
+}
+
+impl IrqAllocator {
+    pub fn new(last_used_irq: u32, max: u32) -> Result<Self> {
+        if last_used_irq >= max {
+            return Err(Error::InvalidValue);
+        }
+        Ok(IrqAllocator { last_used_irq, max })
+    }
+
+    pub fn next_irq(&mut self) -> Result<u32> {
+        match self.last_used_irq.checked_add(1) {
```

Comment:
You can write this in a more Rust-y way by making use of `ok_or` which converts the checked_add result from an Option to an Error.